### PR TITLE
setuptools plugin solution for variables

### LINKED ISSUE #1085: transition to plugins for dedupe variables.
newer versions of pip/setuptools are now confused by the [old-style namespace packages](https://github.com/pypa/setuptools/issues/3537#issuecomment-1218204087).

i previously tried to migrate to the implicit style namespace packages, but the problem with that i don't think there's a way to do that this not backwards incompatible. 

specifically, i don't see how you could use the implicit style namespaces and have something like `from dedupe import Dedupe` work. 

you could do `from dedupe.api import Dedupe` but that's not what we have guaranteed.

so, then maybe we have to move to the [src-layout](https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#src-layout)

### PR DESCRIPTION
**Pull request by fgregg**
Using [setuptools plugin facilities](https://setuptools.pypa.io/en/latest/userguide/entry_point.html#entry-points-for-plugins) to set up variables. will close #1085 

## external plugins
external dedupe-variable plugins will look like https://github.com/dedupeio/dedupe-variable-datetime

in particular, they will have this in their pyproject.toml (or equivalent setup.cfg/setup.py)

```toml
[project.entry-points]
dedupevariables = {datetimetype = "datetimetype:DateTimeType"}
```
https://github.com/dedupeio/dedupe-variable-datetime/blob/fcd36afd000641168d9ae369623df866eeac35f9/pyproject.toml#L16

## to do

- [x] https://github.com/dedupeio/dedupe-variable-datetime
- [ ] https://github.com/dedupeio/dedupe-variable-name
- [ ] https://github.com/dedupeio/dedupe-variable-address
- [ ] https://github.com/dedupeio/dedupe-variable-fuzzycategory

### CODE REVIEW

File: dedupe/datamodel.py [hunk 1]
```diff
@@ -145,6 +160,11 @@ def __setstate__(self, d):
 def typify_variables(
     variable_definitions: Iterable[VariableDefinition],
 ) -> tuple[list[FieldVariable], list[Variable]]:
+
+    variable_types = {}
+    for variable_type in pm.hook.register_variable():
+        variable_types.update(variable_type)
```
Comment (fgregg) [inline dedupe/datamodel.py:166]: would be good to do some protocol checking here.

**Comment by fgregg** [comment]
@NickCrews so here's an implementation inspired by how datasette uses pluggy. would appreciate your thoughts on this.

**Comment by NickCrews** [comment]
This looks good on first passover. But unfortunately since you just invested a bunch of time, I think pluggy is actually overkill (it was my suggestion originally, sorry!!!). See alternate implementation: https://github.com/dedupeio/dedupe/pull/1122

**Comment by fgregg** [comment]
there are reasonably test pattern, 

but i feel you on the boilerplate. it’s about the same amount and much *clearer* boilerplate than the namespace solution.

hmm…

**Comment by fgregg** [comment]
okay, i'm liking this solution a lot better than the pluggy one. thanks for pushing against that @NickCrews!

i'm going to try this out on a local project that needs a custom variable type and if i still like it, i'll write some tests and bring it in.

**Comment by NickCrews** [comment]
This is much better than pluggy, awesome!

I think the basic philosophical difference between this and #1122 is who has the responsibility of defining the lookup key, and whether namespacing with the package name is important. Here, when the author of plugin A writes
```
[project.entry-points]
dedupevariables = {zipcodetype = "pluginA:ZipCodeVariable"}
```
they decide that the way to get this plugin is via the key "zipcodetype". You don't have to know that it lives inside a package called pluginA. If there is another plugin B that wants to do basically the same thing, they can't say `dedupevariables = {zipcodetype = "pluginB:ZipCodeVariable"}` because the key zipcodetype is now ambiguous, do you mean from pluginA or pluginB?

Whereas in #1122, the name of the package, pluginA vs pluginB, is significant, and is used to namespace. Now you can have two different plugins that export variables of the same name. On PyPI pluginA and pluginB are already forced to be using different names, so we can piggyback off that guarantee and now you can't have accidental collisions.

What do you think, are you trying to avoid having to specify "pluginA" vs "pluginB" in your variable definition dicts? I think that is silly, you already know where ZipCodeVariable is coming from, you have the package name written down in your requirements.txt. Am I missing something here? Note this is different from wanting to not care about WHAT FILE in the plugin , which I agree with the user shouldn't have to specify. But plugin authors shoud get around that with what I mention in https://github.com/dedupeio/dedupe/pull/1122#issuecomment-1345801485

File: dedupe/datamodel.py [hunk 2]
```diff
@@ -153,7 +153,10 @@ def typify_variables(
     variable_definitions: Iterable[VariableDefinition],
 ) -> tuple[list[FieldVariable], list[Variable]]:
 
-    variable_types = {variable.type: variable for variable in dedupe.variables.__all__}
+    variable_types = {}
+    for variablename in dedupe.variables.__all__:
```
[the remark this answers is not in the document]
Comment (fgregg) [reply dedupe/datamodel.py:157]: it's intended use is to control the behavior of "import *" but i think this works fine.

**Comment by fgregg** [comment]
i think my main reason for preferring this is consistency of interface. i don't like that there are some variables that we refer to with a short alias and others that we use a path in the datamodel.

we could have a deprecation path where the paths were *the* way to reference variables, but honestly i think that a lot of new people would be confused by that.

if we were going to go down that road, i might even prefer a datamodel that went from:

```python
data_model = [{field: "name", type: "String"},
              {field: "address", type: "String"}]
```

to 

```python
from dedupe.variables import StringVariable
data_model = [StringVariable(field="name"), StringVariable(field="address")]
```

which would avoid the plugin question altogether.

i kind of like that solution, but it seems like a lot of disruption without a huge benefit to users.

**Comment by fgregg** [comment]
closed by #1193

**Pull request closed without merging**